stargz/verify/verify.go

```diff
@@ -120,70 +122,157 @@ func NewVerifiableStagz(sgz *io.SectionReader) (newSgz io.Reader, jtocDigest str
+		io.LimitReader(sgz, toc.offset), // Original stargz (before TOC JSON)
+		pr,                              // Rewritten TOC JSON
+		toc.footer,                      // Unmodified footer (because tocOffset is unchanged)
 	), fmt.Sprintf("sha256:%x", tocHash.Sum(nil)), nil
 }
 
+// StargzTOC checks that the TOC JSON in the passed stargz blob matches the
+// passed digests and that the TOC JSON contains digests for all chunks
+// contained in the stargz blob. If the verification succceeds, this function
+// returns TOCEntryVerifier which holds all chunk digests in the stargz blob.
+func StargzTOC(sgz *io.SectionReader, tocDigest string) (TOCEntryVerifier, error) {
+	toc, err := extractTOCJSON(sgz)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to extract TOC JSON")
+	}
+
+	// Verify the digest of TOC JSON
+	if toc.digest != tocDigest {
+		return nil, fmt.Errorf("invalid TOC JSON %q; want %q", toc.digest, tocDigest)
+	}
+
+	// Collect all chunks digests in this layer
+	digestMap := make(map[int64]digest.Digest) // map from chunk offset to the digest
+	for _, e := range toc.jtoc.Entries {
+		if e.Type == "reg" || e.Type == "chunk" {
+			if e.Type == "reg" && e.Size == 0 {
+				continue // ignores empty file
+			}
+
+			// offset must be unique in stargz blob
+			if _, ok := digestMap[e.Offset]; ok {
+				return nil, fmt.Errorf("offset %d found twice", e.Offset)
+			}
+
+			// all chunk entries must contain digest
+			if e.ChunkDigest == "" {
+				return nil, fmt.Errorf("ChunkDigest of %q(off=%d) not found in TOC JSON",
+					e.Name, e.Offset)
+			}
+
+			d, err := digest.Parse(e.ChunkDigest)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to parse digest %q", e.ChunkDigest)
+			}
+			digestMap[e.Offset] = d
+		}
+	}
+
+	return &verifier{digestMap: digestMap}, nil
+}
+
+// TOCEntryVerifier holds verifiers that are usable for verifying chunks contained
+// in a stargz blob.
+type TOCEntryVerifier interface {
+
+	// Verifier provides a content verifier that can be used for verifying the
+	// contents of the specified TOCEntry.
+	Verifier(ce *stargz.TOCEntry) (digest.Verifier, error)
+}
+
+// verifier is an implementation of TOCEntryVerifier which holds verifiers keyed by
+// offset of the chunk.
+type verifier struct {
+	digestMap   map[int64]digest.Digest
+	digestMapMu sync.Mutex
+}
+
+// Verifier returns a content verifier specified by TOCEntry.
+func (v *verifier) Verifier(ce *stargz.TOCEntry) (digest.Verifier, error) {
+	v.digestMapMu.Lock()
+	defer v.digestMapMu.Unlock()
+	d, ok := v.digestMap[ce.Offset]
+	if !ok {
+		return nil, fmt.Errorf("verifier for offset=%d,size=%d hasn't been registered",
+			ce.Offset, ce.ChunkSize)
+	}
+	return d.Verifier(), nil
+}
+
+// tocInfo is a set of information related to footer and TOC JSON in a stargz
+type tocInfo struct {
+	jtoc   *jtoc         // decoded TOC JSON
+	digest string        // SHA256 digest of TOC JSON. formed as "sha256:abcd1234..."
```

Comment:
consider using this type: https://pkg.go.dev/github.com/opencontainers/go-digest?tab=doc#Digest

Reply:
Fixed patches to use `digest.Digest` instead of `string`.